// app/helpers/WeatherHelper.js
var request = require('request');


var helper = {

    fetchUrl: function(url, callback){
        var timeoutInMilliseconds = 10*1000;
        var opts = {
            url: url,
            timeout: timeoutInMilliseconds
        };

        request(opts, function (err, response, body) {
            if (err) {
                console.log("Error requesting " + url + ": " + err);
                if (err){
                    callback(err, null);
                }
            }
            else{
                console.log(response);
                console.log(body);
                callback(null, body);
            }
        });
    },
    currentWeatherForZipcode : function(zipcode, callback){
        var url = "https://api.apixu.com/v1/current.json?key=" + process.env.APIXU_API_KEY + "&q=" + zipcode;
        this.fetchUrl(url, callback);
    },
    forecastForZipcode : function(zipcode, callback){
        var url = "https://api.apixu.com/v1/forecast.json?key=" + process.env.APIXU_API_KEY + "&q=" + zipcode;
        this.fetchUrl(url, callback);
    }

};


module.exports = helper;
