// app/db/Forecast.js
var mongoose = require('app/db/mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;


var customSchema = new Schema({
        condition: {type: String},
        max_temp_f: {type: String},
        max_temp_c : {type: String},
        min_temp_f : {type: String},
        min_temp_c : {type: String},
        avg_temp_f : {type: String},
        avg_temp_c : {type: String},
        will_it_rain : {type: Number},
        will_it_snow : {type: Number},
        rain_start_time: {type: String},
        rain_end_time: {type: String},
        snow_start_time: {type: String},
        snow_end_time: {type: String},
        is_day : {type: Number},
        icon : {type: String},
        humidity: {type: Number}
    },
    {timestamps: {createdAt: 'dateCreated', updatedAt: 'dateUpdated'}}
);

customSchema.virtual('').get(function () {
});


customSchema.pre("save", function (next) {
    next();
});

customSchema.post("init", function (doc) {

});

customSchema.statics.forecastWithData = function(data, callback){

    var forecast = new this({

    });

    forecast.save(function(err, _forecast){
       if(err){
           console.log(err);
           callback(err, null);
       }
        else{
           console.log("created forecast: " + _forecast);
           callback(null, _forecast);
       }
    });
};


module.exports = mongoose.model('Forecast', customSchema);
