// app/db/HourlyForecast.js
var mongoose = require('app/db/mongoose_connect.js');
var Schema = mongoose.Schema;
var ObjectId = Schema.Types.ObjectId;

var Globals = require('app/helpers/Globals.js');

var customSchema = new Schema({
        condition: {type: String},
        will_it_rain : {type: Number},
        will_it_snow : {type: Number},
        rain_start_time: {type: String},
        rain_end_time: {type: String},
        snow_start_time: {type: String},
        snow_end_time: {type: String},
        is_day : {type: Number},
        icon : {type: String},
        humidity: {type: Number},
        temp_c : {type: Number},
        temp_f : {type: Number},
        wind_mph : {type: Number},
        wind_kmph : {type: Number},
        precip_mm : {type: Number},
        precip_in : {type: Number},
        feelslike_c : {type: Number},
        feelslike_f : {type: Number},
        windchill_c : {type: Number},
        windchill_f : {type: Number}
    },
    {timestamps: {createdAt: 'dateCreated', updatedAt: 'dateUpdated'}}
);

customSchema.virtual('').get(function () {
});

customSchema.statics.withJson = function(json){
  var forecast = new this({
      will_it_rain : json["will_it_rain"],
      will_it_snow : json["will_it_snow"],
      is_day : json["is_day"],
      icon : json["condition"]["icon"],
      temp_f : json["temp_f"],
      temp_c : json["temp_c"],
      windchill_c : json["windchill_c"],
      windchill_f : json["windchill_f"],
      feelslike_c : json["feelslike_c"],
      feelslike_f : json["feelslike_f"],
      precip_mm : json["precip_mm"],
      precip_in : json["precip_in"],
      wind_mph : json["wind_mph"],
      wind_kmph : json["wind_kmph"],
      humidity : json["humidity"]
  });

    return forecast;
};

customSchema.pre("save", function (next) {
    next();
});

customSchema.post("init", function (doc) {

});

module.exports = mongoose.model('HourlyForecast', customSchema);
